### PR TITLE
sql: add doc for split table/index regions

### PR DESCRIPTION
**Pull request by crazycs520**
add doc for `split table/index regions`.

### CODE REVIEW

File: dev/reference/sql/statements/split-region.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
+# Split region 使用文档
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]: Please add the metadata at the beginning of the file, and add this file to TOC.md.

File: dev/reference/sql/statements/split-region.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+t[table_id]_i[index_id][index_value]
+eg
+t22_i5abc # table_id 是22, index_id 是5， index_value 是 abc 
+```
+
+同一表中的同一索引数据的 table_id 和 index_id 是一样的，所以切分索引 region 要根据 index_value 来切分region。
+
+#### 均匀切分
+
+索引均匀切分的与行数据的均匀切分的原理一样，只是计算 step 的值复杂一些，因为 index_value 可能不是整数。 `upper` 和 `lower` value 会先 encode 成byte 数组，去掉 `lower` 和 `upper` byte 数组的最长公共前缀后，从 `lower`, `upper` 各取前8字节转成uint64, 再去计算 `step = ( upper - lower)/num`。计算出 step 后再把 step 给 encode 成byte 数组，添加到之前 `upper`和 `lower`的最长公共前缀后面组成一个 key 后去做切分。下面是示例：
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
索引均匀切分与行数据均匀切分的原理一样，只是计算 step 的值复杂一些，因为 `index_value` 可能不是整数。 `upper` 和 `lower` 的值会先编码成 byte 数组，去掉 `lower` 和 `upper` byte 数组的最长公共前缀后，从 `lower`, `upper` 各取前 8 字节转成 uint64，再去计算 `step = (upper - lower)/num`。计算出 step 后再将 step 编码成 byte
数组，添加到之前 `upper`和 `lower`的最长公共前缀后面组成一个 key 后去做切分。示例如下：
```

File: dev/reference/sql/statements/split-region.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+如果索引 idx 的列也是整数类型，可以用下面 SQL 切分索引数据
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
如果索引 idx 的列也是整数类型，可以用如下 SQL 语句切分索引数据：
```

File: dev/reference/sql/statements/split-region.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+```sql
+SPLIT TABLE t INDEX idx BETWEEN (-9223372036854775808) AND (9223372036854775807) REGIONS 16;
+```
+
+上面SQL 会把表 t 中 idx 索引数据region 从 从minInt64 到 maxInt64 之间均匀切割出16个 region。
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
该语句会把表 t 中 idx 索引数据 region 从 `minInt64` 到 `maxInt64` 之间均匀切割出 16 个 region。
```

File: dev/reference/sql/statements/split-region.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+如果索引 idx1 的列是 varchar 类型，希望根据前缀字母来切分索引数据
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
如果索引 idx1 的列是 varchar 类型，希望根据前缀字母来切分索引数据：
```

File: dev/reference/sql/statements/split-region.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+```sql
+SPLIT TABLE t INDEX idx1 BETWEEN ("a") AND ("z") REGIONS 26;
+```
+
+上面SQL会把表 t 中 idx1 索引数据的region 从 a ~ z 切成 26 个 region，region1 的范围是 [minIndexValue , b), region 2 的范围是 [b , c) ...... region 26 的范围是 [y , minIndexValue] 。对于 idx 索引以 a 为前缀的数据都会写到 region1, 以 b 为前缀的索引数据都会写到 region 2，以此类推。
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
该 SQL 语句会把表 t 中 idx1 索引数据的 region 从 a~z 切成 26 个 region，region1 的范围是 [minIndexValue, b)，region 2 的范围是 [b, c)...... region 26 的范围是 [y, minIndexValue]。对于 idx 索引以 a 为前缀的数据都会写到 region1，以 b 为前缀的索引数据都会写到 region 2，以此类推。
```

File: dev/reference/sql/statements/split-region.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+如果索引 idx2 的列是 timestamp/datatime 等时间类型，希望根据时间区间来切分索引数据
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
如果索引 idx2 的列是 timestamp/datatime 等时间类型，希望根据时间区间来切分索引数据：
```

File: dev/reference/sql/statements/split-region.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+```sql
+SPLIT TABLE t INDEX idx2 BETWEEN ("2010-01-01 00:00:00") AND ("2020-01-01 00:00:00") REGIONS 10;
+
+```
+
+上面SQL 会把表 t 中 idx2 的索引数据 region 从 `2010-01-01 00:00:00` 到  `2020-01-01 00:00:00` 切成 10 个 region。region1 的范围是从` [minIndexValue,  2011-01-01 00:00:00 )`， region2 的范围是 `[2011-01-01 00:00:00 , 2012-01-01 00:00:00)` ......
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
该 SQL 语句会把表 t 中 idx2 的索引数据 region 从 `2010-01-01 00:00:00` 到  `2020-01-01 00:00:00` 切成 10 个 region。region1 的范围是从 `[minIndexValue,  2011-01-01 00:00:00)`，region2 的范围是 `[2011-01-01 00:00:00, 2012-01-01 00:00:00)`......
```

File: dev/reference/sql/statements/split-region.md [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+其他索引列类型的切分方法也是类似的。
+
+对于联合索引的数据 region 切分，唯一不同的是指定的值可以指定多个column 的值
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
对于联合索引的数据 region 切分，唯一不同的是指定的值可以指定多个 column 的值。
```

File: dev/reference/sql/statements/split-region.md [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+
+比如对于索引 `idx3 (a, b)`，他包含2列，a 是 timestamp, b 是 int。 如果只想根据 a 列做时间范围的切分，可以用切分单列时间索引的SQL 语句来切分，`lower_value` 和 `upper_velue` 中不指定 b 列的值即可。
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
比如索引 `idx3 (a, b)` 包含 2 列，a 是 timestamp，b 是 int。 如果只想根据 a 列做时间范围的切分，可以用切分单列时间索引的 SQL 语句来切分，`lower_value` 和 `upper_velue` 中不指定 b 列的值即可。
```

File: dev/reference/sql/statements/split-region.md [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+```
+
+如果想在时间相同的情况下，根据 b 列再做一次切分，在切分时指定 b 列的值即可。
+
+```sql
+SPLIT TABLE t INDEX idx3 BETWEEN ("2010-01-01 00:00:00", "a") AND ("2010-01-01 00:00:00", "z") REGIONS 10;
+```
+
+上面 SQL 在a 列时间前缀相同的情况下，根据 b 列的值从 a ~ z 切了 10 个 region。如果指定的 a 列的值不相同，那么 b 列的值可能不会用到。
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
该 SQL 语句在 a 列时间前缀相同的情况下，根据 b 列的值从 a~z 切了 10 个 region。如果指定的 a 列的值不相同，那么 b 列的值可能不会用到。
```

File: dev/reference/sql/statements/split-region.md [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+
+#### 不均匀切分
+
+索引数据也可以根据用用户指定的索引值来做切分。
+假如 idx4 (a,b), a 列是 varchar 类型， b 列是 timestamp 类型。
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion

假如有 idx4 (a,b)，其中 a 列是 varchar 类型， b 列是 timestamp 类型。
```

File: dev/reference/sql/statements/split-region.md [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+
+```sql
+SPLIT TABLE t1 INDEX idx4 ("a", "2000-01-01 00:00:01"),  ("b", "2019-04-17 14:26:19"),  ("c", "");  
+
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
```

File: dev/reference/sql/statements/split-region.md [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+上面 SQL 在a 列时间前缀相同的情况下，根据 b 列的值从 a ~ z 切了 10 个 region。如果指定的 a 列的值不相同，那么 b 列的值可能不会用到。
+
+#### 不均匀切分
+
+索引数据也可以根据用用户指定的索引值来做切分。
+假如 idx4 (a,b), a 列是 varchar 类型， b 列是 timestamp 类型。
+
+```sql
+SPLIT TABLE t1 INDEX idx4 ("a", "2000-01-01 00:00:01"),  ("b", "2019-04-17 14:26:19"),  ("c", "");
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
SPLIT TABLE t1 INDEX idx4 ("a", "2000-01-01 00:00:01"), ("b", "2019-04-17 14:26:19"), ("c", "");  
```

File: dev/reference/sql/statements/split-region.md [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+
+索引数据也可以根据用用户指定的索引值来做切分。
+假如 idx4 (a,b), a 列是 varchar 类型， b 列是 timestamp 类型。
+
+```sql
+SPLIT TABLE t1 INDEX idx4 ("a", "2000-01-01 00:00:01"),  ("b", "2019-04-17 14:26:19"),  ("c", "");  
+
+```
+
+上面 SQL 指定了3个值，会切分出4个region， 每个region 的范围如下。
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
该 SQL 语句指定了 3 个值，会切分出 4 个 region， 每个 region 的范围如下。
```

File: dev/reference/sql/statements/split-region.md [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+```
+region1  [ minIndexValue               , ("a", "2000-01-01 00:00:01"))
+region2  [("a", "2000-01-01 00:00:01") , ("b", "2019-04-17 14:26:19"))
+region3  [("b", "2019-04-17 14:26:19") , ("c", "")                   )
+region4  [("c", "")                    , maxIndexValue               )
+```
+
+## pre_split_regions
+
+使用带有`shard_row_id_bits`的表时，如果希望建表时就做均匀切分 region，可以考虑配合 `pre_split_regions` 一起使用，用来在建表成功后就开始预均匀切分 `2^(pre_split_regions-1)` 个regions。注意：`pre_split_regions` 必须小于等于 `shard_row_id_bits`。
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
使用带有 `shard_row_id_bits` 的表时，如果希望建表时就做均匀切分 region，可以考虑配合 `pre_split_regions` 一起使用，用来在建表成功后就开始预均匀切分 `2^(pre_split_regions-1)` 个 regions。

> **注意：**
>
> `pre_split_regions` 必须小于等于 `shard_row_id_bits`。
```

File: dev/reference/sql/statements/split-region.md [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+
+### 示例
+
+```sql
+create table t (a int, b int,index idx1(a)) shard_row_id_bits = 4 pre_split_regions=3;
+
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
```

File: dev/reference/sql/statements/split-region.md [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+```
+
+上面建表语句在建表后，会对这个表 t 预切分出 4 + 1 个region。4 ( 4=2^(3-1) ) 个 regions 是用来存 table 的行数据的，1 个 region 是用来存 idx1 索引的数据。
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
该语句在建表后，会对这个表 t 预切分出 4 + 1 个 region。4 (4=2^(3-1)) 个 regions 是用来存 table 的行数据的，1 个 region 是用来存 idx1 索引的数据。
```

File: dev/reference/sql/statements/split-region.md [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+
+4个 table region 的范围区间如下：
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
4 个 table region 的范围区间如下：

```

File: dev/reference/sql/statements/split-region.md [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+```
+region1:   [ -inf      ,  1<<61 )  
+region2:   [ 1<<61     ,  2<<61 )
+region3:   [ 2<<61     ,  3<<61 )
+region4:   [ 3<<61     ,  +inf  )
+```
+
+关于为什么是 切割 2^(pre_split_regions-1) 个regions，因为使用 shard_row_id_bits 时，只会分配正数给 _tidb_rowid，所以就没有必要给负数的那段区间做 split region 了。
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
关于为什么是切割 2^(pre_split_regions-1) 个 regions，因为使用 shard_row_id_bits 时，只会分配正数给 `_tidb_rowid`，所以就没有必要给负数的那段区间做 split region 了。
```

**Comment by yikeke** [comment]
Please address comments and ask for a technical review, thanks! @crazycs520

**Comment by yikeke** [comment]
**Note:** I have updated all code blocks according to our new code block format rules, please see [Update code block format to better support copy](https://internal.pingcap.net/jira/browse/ICM-95) for details.

Content LGTM now. Please ask someone to take a technical review, thanks! @crazycs520

File: dev/reference/sql/statements/split-region.md [hunk 21]
```diff
@@ -0,0 +1,205 @@
+---
```
Comment (yikeke) [inline dev/reference/sql/statements/split-region.md]: Last one: please help add this file into TOC.md.

File: dev/reference/configuration/tidb-server/tidb-specific-variables.md [hunk 22]
```diff
@@ -468,3 +468,22 @@ set tidb_slow_log_threshold = 200
...
+
+### tidb_wait_split_region_finish
+
+作用域：SESSION
+
+默认值：1
+
+由于打散 region 的时间可能比较长，主要由 PD 调度以及 TiKV 的负载情况所决定。这个变量用来设置在执行 `SPLIT REGION` 语句时，是否同步等待所有 region 都打散完成后再返回结果给客户端。默认 1 代表等待打散完成后再返回结果。0 代表不等待 region 打散完成就返回。
+
+需要注意的是，在 region 打散期间，对正在打散 region 上的写入和读取的性能会有一定影响，对比批量写入，导数据等场景，还是建议等待 region 打散完成后再开始导数据。
```
Comment (winkyao) [inline dev/reference/configuration/tidb-server/tidb-specific-variables.md]: 对比 -> 对于

File: dev/reference/sql/statements/split-region.md [hunk 23]
```diff
@@ -0,0 +1,225 @@
...
+region2:   [ 1<<61     ,  2<<61 )
+region3:   [ 2<<61     ,  3<<61 )
+region4:   [ 3<<61     ,  +inf  )
+```
+
+关于为什么是切割 2^(pre_split_regions-1) 个 regions，因为使用 shard_row_id_bits 时，只会分配正数给 `_tidb_rowid`，所以就没有必要给负数的那段区间做 split region 了。
+
+## 相关 session variable
+
+### tidb_wait_split_region_finish
```
Comment (zimulala) [inline dev/reference/sql/statements/split-region.md]: 能直接用连接吗？比如`详细请参考...`。不然 line209-line225 有点重复。

File: dev/reference/sql/statements/split-region.md [hunk 24]
```diff
@@ -0,0 +1,225 @@
...
+
+上述情况中，如果在新建的表上发生大批量写入，则会造成热点，因为开始只有一个 region，所有的写请求都发生在该 region 所在的那台 TiKV 上。
+
+为解决上述场景中的热点问题，TiDB 引入了预切分 region 的功能，即可以根据用户指定的参数，预先为某个表切分出多个 region，并打散到各个 TiKV 上去。
+
+## Split region 的使用
+
+Split region 有 2 种不同的语法，具体如下：
+
+{{< copyable "sql" >}}
```
Comment (zimulala) [inline dev/reference/sql/statements/split-region.md]: 这个是什么？

Comment (yikeke) [reply dev/reference/sql/statements/split-region.md]: 是新的代码块写法，为了让下面的代码块内容能够复制，详见 [Update code block format to better support copy](https://internal.pingcap.net/jira/browse/ICM-95)。

Comment (zimulala) [reply dev/reference/sql/statements/split-region.md]: 好的，知道了

File: dev/reference/sql/statements/split-region.md [hunk 25]
```diff
@@ -0,0 +1,225 @@
...
+该语句指定了 3 个值，会切分出 4 个 region，每个 region 的范围如下。
+
+```
+region1  [ minIndexValue               , ("a", "2000-01-01 00:00:01"))
+region2  [("a", "2000-01-01 00:00:01") , ("b", "2019-04-17 14:26:19"))
+region3  [("b", "2019-04-17 14:26:19") , ("c", "")                   )
+region4  [("c", "")                    , maxIndexValue               )
+```
+
+## pre_split_regions
```
Comment (zimulala) [inline dev/reference/sql/statements/split-region.md]: 能将 `create table` 里面这个描述连接到这里吗？

File: dev/reference/sql/statements/split-region.md (hunk 23 above)
Comment (crazycs520) [reply dev/reference/sql/statements/split-region.md]: done.

File: dev/reference/sql/statements/split-region.md (hunk 25 above)
Comment (crazycs520) [reply dev/reference/sql/statements/split-region.md]: 你直接加的 描述是被删除了吗？找不到了

File: dev/reference/sql/statements/split-region.md [hunk 26]
```diff
@@ -0,0 +1,209 @@
...
+
+```go
+t22_r11
+```
+
+同一表中行数据的 `table_id` 是一样的，但 `row_id` 肯定不一样，所以可以根据 `row_id` 来切分 region。
+
+#### 均匀切分
+
+由于 `row_id` 是整数，所以根据用户指定的 `lower_value`、`upper_value` 以及 `region_num`，可以很容易地计算出需要切分的 key。先计算出 step（`step = (upper_value - lower)/num`），然后在 `lower` 和 `upper` 之间每隔 step 区间切一次，最终切出 `num` 个 region。
```
Comment (tangenta) [inline dev/reference/sql/statements/split-region.md]: (upper_value - lower) -> (upper_value - lower_value)
or
(upper_value - lower) -> (upper - lower) ?

File: dev/reference/sql/statements/split-region.md [hunk 27]
```diff
@@ -0,0 +1,209 @@
...
+
+{{< copyable "sql" >}}
+
+```sql
+SPLIT TABLE t INDEX idx1 BETWEEN ("a") AND ("z") REGIONS 26;
+```
+
+该语句会把表 t 中 idx1 索引数据的 region 从 a
+
+
```
Comment (tangenta) [inline dev/reference/sql/statements/split-region.md]: Why are there two blank lines here?

File: dev/reference/sql/statements/split-region.md [hunk 28]
```diff
@@ -0,0 +1,209 @@
...
+z 切成 26 个 region，region1 的范围是 [minIndexValue, b)，region 2 的范围是 [b, c)，……，region 26 的范围是 [y, minIndexValue]。对于 idx 索引以 a 为前缀的数据都会写到 region1，以 b 为前缀的索引数据都会写到 region 2，以此类推。
```
Comment (tangenta) [inline dev/reference/sql/statements/split-region.md]:
```suggestion
z 切成 26 个 region，region1 的范围是 [minIndexValue, b)，region 2 的范围是 [b, c)，……，region 26 的范围是 [y, minIndexValue]。对于 idx 索引以 a 为前缀的数据都会写到 region 1，以 b 为前缀的索引数据都会写到 region 2，以此类推。
```

File: dev/reference/sql/statements/split-region.md [hunk 29]
```diff
@@ -0,0 +1,209 @@
...
+
+`BETWEEN lower_value AND upper_value REGIONS region_num` 语法是通过指定上、下边界和 region 数量，然后在上、下边界之间均匀切分出 `region_num` 个 region。
+
+{{< copyable "sql" >}}
+
+```sql
+SPLIT TABLE table_name [INDEX index_name] BY (value_list) [, (value_list)] ...
+```
+
+`BY value_list…` 语法将手动指定一系列的点，然后根据这些指定的点切分 region，适用于数据不均匀分布的场景。
```
Comment (tangenta) [inline dev/reference/sql/statements/split-region.md]: IMHO, 不均匀分布 -> 分布不均匀？

File: dev/reference/sql/statements/split-region.md (hunk 28 above)
Comment (tangenta) [reply dev/reference/sql/statements/split-region.md]: Maybe there are only 25 regions here, not 26..?

File: dev/reference/sql/statements/split-region.md (hunk 27 above)
Comment (yikeke) [reply dev/reference/sql/statements/split-region.md]: I fixed this in another PR https://github.com/pingcap/docs-cn/pull/1487.

File: dev/reference/sql/statements/split-region.md (hunk 26 above)
Comment (yikeke) [reply dev/reference/sql/statements/split-region.md]: PTAL. @crazycs520 If there's a mistake, please pull another PR to fix, thanks!

Comment (crazycs520) [reply dev/reference/sql/statements/split-region.md]: fix in https://github.com/pingcap/docs-cn/pull/1509

File: dev/reference/sql/statements/split-region.md (hunk 28 above)
Comment (crazycs520) [reply dev/reference/sql/statements/split-region.md]: No, It should be 26 regions. But `region 26 is [z, maxIndexValue)`

**Pull request merged**